### PR TITLE
Clean deprecated metrics

### PR DESCRIPTION
**Pull request by danielqsj**
**What type of PR is this?**

/kind cleanup
/sig instrumentation

**What this PR does / why we need it**:

As the deprecation plan of [kubernetes metrics overhaul](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0031-kubernetes-metrics-overhaul.md#deprecation-plan), we should remove the deprecated metrics in v1.18.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
The following deprecated metrics are removed, please convert to the corresponding metrics:
1. The following replacement metrics are available from v1.14.0:
* `rest_client_request_latency_seconds` -> `rest_client_request_duration_seconds`
* `scheduler_scheduling_latency_seconds` -> `scheduler_scheduling_duration_seconds `
* `docker_operations` -> `docker_operations_total`
* `docker_operations_latency_microseconds` -> `docker_operations_duration_seconds`
* `docker_operations_errors` -> `docker_operations_errors_total`
* `docker_operations_timeout` -> `docker_operations_timeout_total`
* `network_plugin_operations_latency_microseconds` -> `network_plugin_operations_duration_seconds`
* `kubelet_pod_worker_latency_microseconds` -> `kubelet_pod_worker_duration_seconds`
* `kubelet_pod_start_latency_microseconds` -> `kubelet_pod_start_duration_seconds`
* `kubelet_cgroup_manager_latency_microseconds` -> `kubelet_cgroup_manager_duration_seconds`
* `kubelet_pod_worker_start_latency_microseconds` -> `kubelet_pod_worker_start_duration_seconds`
* `kubelet_pleg_relist_latency_microseconds` -> `kubelet_pleg_relist_duration_seconds`
* `kubelet_pleg_relist_interval_microseconds` -> `kubelet_pleg_relist_interval_seconds`
* `kubelet_eviction_stats_age_microseconds` -> `kubelet_eviction_stats_age_seconds`
* `kubelet_runtime_operations` -> `kubelet_runtime_operations_total`
* `kubelet_runtime_operations_latency_microseconds` -> `kubelet_runtime_operations_duration_seconds`
* `kubelet_runtime_operations_errors` -> `kubelet_runtime_operations_errors_total`
* `kubelet_device_plugin_registration_count` -> `kubelet_device_plugin_registration_total`
* `kubelet_device_plugin_alloc_latency_microseconds` -> `kubelet_device_plugin_alloc_duration_seconds`
* `scheduler_e2e_scheduling_latency_microseconds` -> `scheduler_e2e_scheduling_duration_seconds`
* `scheduler_scheduling_algorithm_latency_microseconds` -> `scheduler_scheduling_algorithm_duration_seconds`
* `scheduler_scheduling_algorithm_predicate_evaluation` -> `scheduler_scheduling_algorithm_predicate_evaluation_seconds`
* `scheduler_scheduling_algorithm_priority_evaluation` -> `scheduler_scheduling_algorithm_priority_evaluation_seconds`
* `scheduler_scheduling_algorithm_preemption_evaluation` -> `scheduler_scheduling_algorithm_preemption_evaluation_seconds`
* `scheduler_binding_latency_microseconds` -> `scheduler_binding_duration_seconds`
* `kubeproxy_sync_proxy_rules_latency_microseconds` -> `kubeproxy_sync_proxy_rules_duration_seconds`
* `apiserver_request_latencies` -> `apiserver_request_duration_seconds`
* `apiserver_dropped_requests` -> `apiserver_dropped_requests_total`
* `etcd_request_latencies_summary` -> `etcd_request_duration_seconds`
* `apiserver_storage_transformation_latencies_microseconds ` -> `apiserver_storage_transformation_duration_seconds`
* `apiserver_storage_data_key_generation_latencies_microseconds` -> `apiserver_storage_data_key_generation_duration_seconds`
* `apiserver_request_count` -> `apiserver_request_total`
* `apiserver_request_latencies_summary`
2. The following replacement metrics are available from v1.15.0:
* `apiserver_storage_transformation_failures_total` -> `apiserver_storage_transformation_operations_total`
```

### CODE REVIEW

**Comment by brancz** [comment]
I'm happy with having this PR out for now, but I'd like to hold off on merging it until we adapt some of the downstream usages as we promised in the KEP. Such as https://github.com/kubernetes-monitoring/kubernetes-mixin/issues/182

**Comment by danielqsj** [comment]
@brancz sure. I will take time to look at kubernetes-monitoring/kubernetes-mixin, and see how to help it.

/retest

**Comment by danielqsj** [comment]
@brancz I saw the ticket https://github.com/kubernetes-monitoring/kubernetes-mixin/issues/182 has been closed. Is it time to move on?

**Comment by brancz** [comment]
Yes we can move forward with this.

/lgtm

**Comment by danielqsj** [comment]
thanks @brancz 

/cc @liggitt @smarterclayton for approve

**Comment by logicalhan** [comment]
We should get approvals from each corresponding sig owners before letting this through, since it gets rid of many things and therefore has large potential downstream impact. 

/hold

**Comment by danielqsj** [comment]
As the deprecation plan of [kubernetes metrics overhaul](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0031-kubernetes-metrics-overhaul.md#deprecation-plan), we should remove the deprecated metrics in v1.15.

Need your approval from each SIGs:
- `apiserver` and `client-go` part: @lavalamp @smarterclayton @deads2k @sttts @liggitt
- `kubelet` part: @dchen1107 @derekwaynecarr
- `scheduling` part: @bsalamat @k82cn
- `network` part: @thockin @caseydavenport @dcbw

Sorry to bother all of you, but it's more safe to let you know about these changes.

/remove-priority important-longterm
/priority important-soon

**Comment by k82cn** [comment]
I'd suggest to also send an email to k/user & k/dev again to make sure everyone aware this change.

/approve for scheduling part.

**Comment by danielqsj** [comment]
thanks @k82cn . 
I have sent an email to k/devs to inform about this change: https://groups.google.com/forum/#!topic/kubernetes-dev/sxO3CCNx5mU .

**Comment by danielqsj** [comment]
The time is close to the v1.15 code freeze (Thursday, May 30th, 2019).
Could we move a little faster?

Need your approval from each SIGs:

- apiserver and client-go part: @lavalamp @smarterclayton @deads2k @sttts @liggitt
- kubelet part: @dchen1107 @derekwaynecarr
- network part: @thockin @caseydavenport @dcbw

/remove-priority important-soon
/priority critical-urgent

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: Our scalability tests are still relying on that metric.
Can we please leave that for a bit more time until we migrate tests to it?

[It's a bunch of non-trivial work, that has been started at the beginning of the release, but still is not finished.]

@mm4tt @krzysied @mborsz

**Review by wojtek-t — COMMENTED** [review]
/hold

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (danielqsj) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: after offline discuss with @brancz .
If there is an explicit reason and a timeline to fix the dependencies and eventually remove them, we agree to keep the old ones.

Comment (danielqsj) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: @wojtek-t which metrics are scalability tests relying on ?
Just this one `apiserver_request_latencies_summary` ?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: Yes - we just need `apiserver_request_latencies_summary`.
We're already working on no-longer-relying on those and we have it working. It's mostly ensuring that it is stable, works-as-expected and potentially some bug fixing.
So I really expect we will be able to remove that in 1.16 (hopefully in the relatively early phase of that).

Comment (danielqsj) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: thanks.
Bring it back in https://github.com/kubernetes/kubernetes/pull/76496/commits/2b7353f48008567ce10ea671bef5114ca71de1aa. PTAL

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: Thanks a lot - this looks good to me now.

**Comment by logicalhan** [comment]
@danielqsj, would you mind adding back apiserver_request_count as well? We should be able to remove that one by 1.16, but 1.15 may be difficult for us.

File: pkg/scheduler/metrics/metrics.go
Comment (bsalamat) [inline pkg/scheduler/metrics/metrics.go]: We cannot remove the scheduler metrics. These are needed for backward compatibility. We must keep them for at least a year after deprecation (1.14).

**Review by bsalamat — CHANGES_REQUESTED** [review]

File: pkg/scheduler/metrics/metrics.go
Comment (danielqsj) [reply pkg/scheduler/metrics/metrics.go]: @bsalamat  you mean all scheduler metrics must be kept ?
cc @brancz

**Comment by danielqsj** [comment]
@logicalhan  sure. Add `apiserver_request_count` back for one more release.

**Comment by logicalhan** [comment]
> Han Kang sure. Add `apiserver_request_count` back for one more release.

Thanks!

File: pkg/scheduler/metrics/metrics.go
Comment (brancz) [reply pkg/scheduler/metrics/metrics.go]: Metrics are not covered by stability agreements. Unless there is a pressing issue like the kubernetes scalability tests themselves, I don't think we should be making further exceptions. The intention of changing these metrics was long communicated and so was the removal.

**Comment by danielqsj** [comment]
This should be a target in v1.15.
And already spent a long time for reviewing and informed to all k/dev folks.

If there is no other request to keep some specific metrics for pressing issues, could we move on the merge process?

/assign @smarterclayton @liggitt @lavalamp 
/cc @brancz

**Comment by lavalamp** [comment]
Sorry, this is the first I've seen of this. There's an email thread on this subject that afaict has not converged; I pinged it: https://groups.google.com/d/msg/kubernetes-api-reviewers/kdxON6-TtNE/lwiEPWhwCAAJ

**Comment by brancz** [comment]
Actual removal is planned for 1.18, do we want to tag this PR with that information?

**Comment by danielqsj** [comment]
@brancz I suppose so. Could you help add the milestone label ?

**Comment by RainbowMango** [comment]
@danielqsj 
Just a reminder, you can start this work now, since v1.17 has been released.

**Comment by danwinship** [comment]
/assign @bowei 
(also, needs-rebase)

**Comment by danielqsj** [comment]
thanks @RainbowMango  @danwinship 
I start working on the rebase work

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (danielqsj) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: @logicalhan did this deprecated metrics still need to be kept ?

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (danielqsj) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: @wojtek-t did this deprecated metrics still need to be kept ?

**Comment by danielqsj** [comment]
@brancz @liggitt PR rebased and updated, please take another look, thanks

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: No need to kept.  What does `fixing the dependencies` mean? dependency trouble?

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: Same as above. No need to kept.

Comment (danielqsj) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: ref https://github.com/kubernetes/kubernetes/pull/76496#discussion_r287999059

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (danielqsj) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: ref https://github.com/kubernetes/kubernetes/pull/76496#issuecomment-496565529

Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: > would you mind adding back apiserver_request_count as well? We should be able to remove that one by 1.16, but 1.15 may be difficult for us.

We are in 1.17 now. So it's safe to remove them.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: Seems we still need to hold it, as far as I know, perf test needs it. 
But left it to @wojtek-t to confirm.

Comment (mm4tt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: Perf test doesn't need it anymore, we migrated off this metric in https://github.com/kubernetes/kubernetes/issues/73519
/lgtm from the SIG Scalability

Comment (danielqsj) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: thanks @mm4tt , will remove it later

**Comment by brancz** [comment]
These were meant to be removed for 1.18, so we can move ahead with this no?

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (danielqsj) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: cleaned

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (danielqsj) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: cleaned

**Comment by danielqsj** [comment]
@brancz rebased and cleaned all metrics needed.

**Comment by danielqsj** [comment]
friendly ping maintainers @liggitt @smarterclayton @thockin for reviewing

**Comment by liggitt** [comment]
it would be helpful to note in the release note what releases the replacement metrics are available from

**Comment by brancz** [comment]
Agreed. All but `transformation_failures_total` were deprecated in 1.14. @danielqsj could you add that to the release notes?

**Comment by danielqsj** [comment]
@liggitt @brancz agree for your suggestion. Release notes are updated.
Please take another look, thanks.

**Comment by brancz** [comment]
I think we're ready to move ahead with this.

/hold cancel

**Comment by coderanger** [comment]
@danielqsj It appears to have been cut off in the quoted patch notes but the new name for `apiserver_dropped_requests_total` is `apiserver_dropped_requests_total` I think?

**Comment by danielqsj** [comment]
@coderanger the summary metrics is deprecated without new replace one. We can use the histograms metrics instead `apiserver_request_duration_seconds`.

ref https://github.com/kubernetes/kubernetes/pull/72336#discussion_r252440578

**Pull request merged**
